libyul/AST.h

```diff
@@ -35,54 +35,62 @@ namespace solidity::yul
 /// Block-scope variable declaration ("let x:u256 := mload(20:u256)"), non-hoisted
-struct VariableDeclaration { langutil::SourceLocation location; TypedNameList variables; std::unique_ptr<Expression> value; };
+struct VariableDeclaration { std::shared_ptr<DebugData> debugData; TypedNameList variables; std::unique_ptr<Expression> value; };
 /// Block that creates a scope (frees declared stack variables)
-struct Block { langutil::SourceLocation location; std::vector<Statement> statements; };
+struct Block { std::shared_ptr<DebugData> debugData; std::vector<Statement> statements; };
 /// Function definition ("function f(a, b) -> (d, e) { ... }")
-struct FunctionDefinition { langutil::SourceLocation location; YulString name; TypedNameList parameters; TypedNameList returnVariables; Block body; };
+struct FunctionDefinition { std::shared_ptr<DebugData> debugData; YulString name; TypedNameList parameters; TypedNameList returnVariables; Block body; };
 /// Conditional execution without "else" part.
-struct If { langutil::SourceLocation location; std::unique_ptr<Expression> condition; Block body; };
+struct If { std::shared_ptr<DebugData> debugData; std::unique_ptr<Expression> condition; Block body; };
 /// Switch case or default case
-struct Case { langutil::SourceLocation location; std::unique_ptr<Literal> value; Block body; };
+struct Case { std::shared_ptr<DebugData> debugData; std::unique_ptr<Literal> value; Block body; };
 /// Switch statement
-struct Switch { langutil::SourceLocation location; std::unique_ptr<Expression> expression; std::vector<Case> cases; };
-struct ForLoop { langutil::SourceLocation location; Block pre; std::unique_ptr<Expression> condition; Block post; Block body; };
+struct Switch { std::shared_ptr<DebugData> debugData; std::unique_ptr<Expression> expression; std::vector<Case> cases; };
+struct ForLoop { std::shared_ptr<DebugData> debugData; Block pre; std::unique_ptr<Expression> condition; Block post; Block body; };
 /// Break statement (valid within for loop)
-struct Break { langutil::SourceLocation location; };
+struct Break { std::shared_ptr<DebugData> debugData; };
 /// Continue statement (valid within for loop)
-struct Continue { langutil::SourceLocation location; };
+struct Continue { std::shared_ptr<DebugData> debugData; };
 /// Leave statement (valid within function)
-struct Leave { langutil::SourceLocation location; };
+struct Leave { std::shared_ptr<DebugData> debugData; };
 
 struct LocationExtractor
 {
-	template <class T> langutil::SourceLocation operator()(T const& _node) const
+	template <class T> std::shared_ptr<DebugData> operator()(T const& _node) const
 	{
-		return _node.location;
+		return _node.debugData;
 	}
 };
 
 /// Extracts the source location from an inline assembly node.
-template <class T> inline langutil::SourceLocation locationOf(T const& _node)
+template <class T> inline std::shared_ptr<DebugData> locationOf(T const& _node)
```

Comment:
I think this is a bad idea, this should stay at returning `langutil::SourceLocation` (and not the origin one).